### PR TITLE
thorium-browser: init at 120.0.6099.235

### LINKED ISSUE #184823: Thorium Browser
**Project description**
Chromium fork with performance & privacy benefits.

**Metadata**

* homepage URL: https://thorium.rocks/
* source URL: https://github.com/Alex313031/Thorium
* license:  GPL-3.0
* platforms: Linux, Darwin

### PR DESCRIPTION
**Pull request by yliceee**
## Description of changes
Adding [thorium browser](https://github.com/Alex313031/thorium). Closes #184823. Built off of #261718.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by rgri** [comment]
Everything is working except for `thorium_shell`. I can get it running by patching `$out/bin/..thorium-shell-wrapped-wrapped` (replacing `/out` with `$out/out`). But, it's failing with:
```
/nix/store/gw2zv8lcvv1pxzbxia8mkb142dq1d5m8-thorium-browser-120.0.6099.235/bin/..thorium-shell-wrapped-
wrapped: line 3: 169322 Trace/breakpoint trap   (core dumped) /nix/store/gw2zv8lcvv1pxzbxia8mkb142dq1d5
m8-thorium-browser-120.0.6099.235/opt/chromium.org/thorium/thorium_shell --enable-experimental-web-plat
form-features --debug --enable-clear-hevc-for-testing $@
[169325:169325:0100/000000.285630:ERROR:zygote_linux.cc(623)] Zygote could not fork: process_type gpu-p
rocess numfds 3 child_pid -1
[169325:169325:0100/000000.285756:ERROR:zygote_linux.cc(655)] write: Broken pipe (32)
```
So far the only issues I've seen similar to this are from appimages. If anyone has an idea, let me know.

**Comment by KiaraGrouwstra** [comment]
- [x] reviewed the diff and commit messages
- [x] made sure ofBorg build succeeded for all applicable platforms
- [x] run nix-review for a reasonable amount of time without any failures (or marked preexisting failures as broken)

**Comment by KiaraGrouwstra** [comment]
@rgri btw it could be of interest to also incorporate thorium at `programs.chromium` like the others there

**Comment by rgri** [comment]
@KiaraGrouwstra thank you for the suggestion. I also see that there is a generalized test suite for some of the chromium packages, which might be good to hook into.

Do you think it would be appropriate to ping a chromium maintainer for advise on best practices? I don't see anything about this in official manuals.

**Comment by KiaraGrouwstra** [comment]
ya let's try that ^^ @networkException @emilylange

File: pkgs/by-name/th/thorium-browser/package.nix [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+      --replace /usr/bin $out/bin \
+      --replace /opt $out/opt
+    substituteInPlace $out/share/applications/thorium-browser.desktop \
+      --replace /usr/bin $out/bin \
+      --replace StartupWMClass=thorium StartupWMClass=thorium-browser \
+      --replace Icon=thorium-browser Icon=$out/opt/chromium.org/thorium/product_logo_256.png
+    addAutoPatchelfSearchPath $out/chromium.org/thorium
+    addAutoPatchelfSearchPath $out/chromium.org/thorium/lib
+    substituteInPlace $out/opt/chromium.org/thorium/thorium-browser \
+      --replace 'export LD_LIBRARY_PATH' "export LD_LIBRARY_PATH=\$LD_LIBRARY_PATH:${lib.makeLibraryPath buildInputs}:$out/chromium.org/thorium:$out/chromium.org/thorium/lib"
```
Comment (SuperSandro2000) [inline pkgs/by-name/th/thorium-browser/package.nix:153]: Do we really need to set LD_LIBRARY_PATH to everything? Isn't autopatchelfHook taking care of this?

**Comment by JamesMowery** [comment]
Are there any blockers for this?

**Comment by rgri** [comment]
I think all that needs to be done is to investigate Sandro's question. It seems from Emily's super thorough advice that adding Thorium to the test suite is a separate project.

**Comment by ljuzig** [comment]
Hey guys, awesome job! I tested it on my local nixpkgs and everything worked smoothly and I was also able to install version 121 by modifying nix with no issue. However, there is something that seems to have been not taken into account. The author deprecated the old release system introducing a multi-flavor build depending on the compatibility of the cpu instruction sets. From now on there will be builds for SSE3 (like the previous ones), and two more compatibility builds for older processors, namely AVX2 and AVX. Perhaps this should be taken into consideration as of now? For instance, if the flavor were to be included in the package name like thorium-browser-sse3 (which should be v120 release contained in the pr).

**Comment by OEvgeny** [comment]
I also tested the packaged version and it is lacking GPU support on my machine. Tested on AMD Ryzen 5 4600H.

**Comment by ljuzig** [comment]
@OEvgeny I can confirm that GPU acceleration is unavailable.

I tested a custom nix script I made that installs thorium through its .appimage and there it's working:
image

Here's my test script:
```nix
{ pkgs ? import <nixpkgs> {} }:
let
  thoriumVersion = "121.0.6167.204";
  thoriumSrc = {
    x86_64-linux = "https://github.com/Alex313031/thorium/releases/download/M${thoriumVersion}/Thorium_Browser_${thoriumVersion}_SSE3.AppImage";
    aarch64-linux = "https://github.com/Alex313031/Thorium-Raspi/releases/download/M${thoriumVersion}/Thorium_Browser_${thoriumVersion}_arm64.AppImage";
  };
  
  makeThorium = system: let
    name = "thorium";
    version = thoriumVersion;
    src = pkgs.fetchurl {
      url = thoriumSrc.${system};
      sha256 = "1be67ce70edded84ffd37b6a707d553091a87a7a1e81cb6d6f1cf7f2ed0959c2";
    };
    appimageContents = pkgs.appimageTools.extractType2 { inherit name src; };
  in pkgs.appimageTools.wrapType2 {
    inherit name version src;
    extraInstallCommands = ''
      install -Dm444 ${appimageContents}/thorium-browser.desktop $out/share/applications/thorium-browser.desktop
      install -Dm444 ${appimageContents}/thorium.png $out/share/icons/hicolor/512x512/apps/thorium.png
    '';
  };
in
{
  thorium_x86_64-linux = makeThorium "x86_64-linux";
  thorium_aarch64-linux = makeThorium "aarch64-linux";
}
```

note: version 121 or 120 the result is the same on my end.

**Comment by JohnRTitor** [comment]
> My best guess is that the Mesa driver mismatch might be the reason why GPU acceleration doesn't work

Yes that's expected. You'll get errors like this too if you are using Chromium browsers from unstable on stable NixOS system. 

It will be fixed once this is merged and used like a normal package from nixpkgs.

**Comment by samos667** [comment]
Hi what's blocking the init of this package ? I have tested the same implementation as this PR with the most recent version of thorium and that seems really fine. https://github.com/samos667/nixpkgs/commit/a6a59e6eac2882dd0311be031ba8789ae23f606f

People who suffer from the lack of hardware accel support is not the fact of using an Nvidia GPU ? Who's if (for the moment) unsupported by chrom*. Or AMD/Intel users do not have hardware acceleration too ? (So unlike others chrom*, and meaning that's a regression of the fork).

(Sadly, I can't respond to myself as I'm a Nvidia user...)

**Comment by samos667** [comment]
wow t4t !

**Comment by JohnRTitor** [comment]
I merged it, further improvements can come in later PRs.

**Comment by mweinelt** [comment]
```
nix-repl> thorium-browser.version
"120.0.6099.235"

nix-repl> chromium.version
"127.0.6533.119"
```

@JohnRTitor Are you joking? Because I'm not laughing.

File: pkgs/by-name/th/thorium-browser/package.nix [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+  nss,
+  pango,
+  pciutils,
+  pipewire,
+  qt6,
+  speechd,
+  systemd,
+  udev,
+  _7zz,
+  vaapiVdpau,
```
Comment (wegank) [inline pkgs/by-name/th/thorium-browser/package.nix:44]: I guess this should be `libva-vdpau-driver` instead?

Comment (mweinelt) [reply pkgs/by-name/th/thorium-browser/package.nix:44]: Either way, this brok eval.

Comment (JohnRTitor) [reply pkgs/by-name/th/thorium-browser/package.nix:44]: #336138

**Comment by mweinelt** [comment]
It is our opinion, that the security posture of the upstream project is insufficient to expose our users to it.

See also https://github.com/NixOS/nixpkgs/pull/336138#issuecomment-2299603455.

Maybe you can consider chromium or ungoogled-chromium instead.

**Pull request merged**